R/epidist_db.R

```diff
@@ -0,0 +1,72 @@
+#' Creates an `epidist` object directly from the epiparameter library (database)
+#'
+#' @description This function can extract an `epidist` object directly from the
+#' library of epidemiological parameters without having to read in an `epiparam`
+#' object and pull out an `epidist` object from one of the entries. If a
+#' distribution from a specific study is required, the `author` argument can
+#' be specified. For now this must match the author entry in the database
+#' exactly to be recognised.
+#'
+#' @details The current criteria for chosing a distribution when multiple
+#' entries exist in the database for a given disease and epidemiological
+#' distribution is to return the entry with the largest sample size.
+#'
+#' @param disease A character string specifying the disease
+#' @param epi_dist A character string specifying the epidemiological
+#' distribution
+#' @param author The author of the study reporting the distribution
+#'
+#' @return An `epidist` object
+#' @export
+#'
+#' @examples
+#' epidist_db(disease = "influenza", epi_dist = "serial_interval")
+epidist_db <- function(disease,
+                       epi_dist = c("incubation_period",
+                                    "onset_to_admission",
+                                    "onset_to_death",
+                                    "serial_interval",
+                                    "generation_time",
+                                    "offspring_distribution"),
+                       author = NULL) {
+
+  epi_dist <- match.arg(arg = epi_dist, several.ok = FALSE)
+
+  # read in database
+  eparam <- epiparam(epi_dist = epi_dist)
+
+  # match disease names against data
+  disease <- match.arg(
+    arg = disease,
+    choices = unique(eparam$disease),
+    several.ok = FALSE
+  )
+
+  # filter based on pathogen and delay distribution
+  eparam <- eparam[eparam$disease == disease, ]
+
+  # extract study or default to largest sample size
+  if (is.null(author)) {
+    eparam <- eparam[eparam$sample_size == max(eparam$sample_size), ]
+  } else {
+    study <- match.arg(
+      arg = author,
+      choices = unique(eparam$author),
+      several.ok = FALSE
+    )
+    eparam <- eparam[eparam$author == study, ]
+  }
+
+  # Throw warning if more than one row found and select first
+  if (nrow(eparam) > 1) {
+    warning("More than one study found. Selecting first one.
+    Please report an issue with duplicated studies.")
+    eparam <- eparam[1, ]
+  }
+
+  # convert epiparam to epidist
+  edist <- as_epidist(x = eparam)
+
+  # return epidist (or list of epidist)
+  edist
+}
```

Comment:
1. Overall not convinced that this 'extra' way to access the parameter library is required - is `as_epidist()` not sufficient?
2. Suggest function title is an imperative rather than description ("Create ..." rather than "Creates ...").
3. Function is difficult to use, with error messages non-informative. For example, when looking for the onset to death distribution for EVD using `epidist_db(disease = "ebola virus disease", epi_dist = "onset_to_death")`, I get this error: `Error in match.arg(arg = disease, choices = unique(eparam$disease), several.ok = FALSE) : 'arg' should be “marburg virus disease”`, which could be replaced with the more informative "X distribution not availabe for Y disease".
4. Related to (2.) above, has the onset to death distribution for EVD been removed? Why? Note that this will break some workflows for `{datadelay}` (together with removal of encapsulated functions).
5. Trying to create an `epidist` object from the database takes a relativel long time (15-ish seconds) for some diseases (such as dengue; try: `dengue_incubation = epidist_db(disease = "dengue", epi_dist = "incubation_period")`) - I assume this is due to parameters being estimated from summary statistics - suggest printing a message to notify users that this may take a few seconds, otherwise it gives the (wrong) impression that something is broken.

Reply:
It would also be useful to have some examples of interactions with other classes in the documentation for these kinds of functions, e.g. is it possible to show a comparison of pulling out an `epidist` object from one of the database entries vs using `epidist_db`?

Reply:
On point 4, the onset to death distribution for ebola has not been added since the update to the library in `{epiparameter}`. It will be added in the near future.

Reply:
Fair enough, will tag @thimotei here to warn that this affects {datadelay} examples and tests as well.

Reply:
On the first point raised, I initially didn't write this function for those reasons, however, I found the convenience of the original `epidist()` function (which read directly from the library) nice and in the scenario that someone is under a time-pressure and cannot look through the library, `epidist_db()` provides a streamlined function for accessing the library, so I'm going to keep it in the package for a while. If the users provide feedback that it is useless or impedes there ability to access data I'll consider removing.

Reply:
Sounds like a good plan - was not suggesting immediate removal.

Reply:
Updated `epidist_db()` to have a nicer error message and added an example of how different functions/classes can be used to get to the same outcome. Resolved in d5b24f13e6fcfcebc9c61f6c726f752f8696a806